in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+  const decodedReturn = decodeSimulatedTx(simulatedTx, builder.functionAbi);
+  return {
+    tx: simulatedTx.tx,
+    publicOutput: simulatedTx.publicOutput,
+    privateOutput: simulatedTx.privateReturnValues,
+    executionResult: simulatedTx.executionResult,
+    result: decodedReturn,
+    request: userRequest,
+  };
+}
+```
+
+### BaseWallet.simulate
+
+```ts
+
+async simulate(userRequest: UserRequest): {
+    tx: simulatedTx.tx,
+    publicOutput: simulatedTx.publicOutput,
+    privateOutput: simulatedTx.privateReturnValues,
+    executionResult: simulatedTx.executionResult,
+    result: decodedReturn,
+    request: initRequest,
+} {
+  // If we're simulating, we need to have the payment method set.
+  // Users should use `read` if they just want to see the result.
+  if (!userRequest.paymentMethod){
+    throw new Error('Payment method must be set before simulating');
+  }
+
+  const builder = new UserRequestBuilder(userRequest);
+
+  await this.#ensureFunctionAbis(builder);
+
+  if (builder.gasSettings) {
+    return this.#simulateInner(builder.build());
+  }
+
+  // If we're paying, e.g. in bananas, figure out how much AZT that is.
+  // Note: paymentMethod.getEquivalentAztBalance() may call `read` internally.
+  const equivalentAztBalance = await builder.paymentMethod.getEquivalentAztBalance();
+  gasEstimator = new BinarySearchGasEstimator(equivalentAztBalance);
+  builder.setGasSettings(gasEstimator.proposeGasSettings());
+
+  while (!gasEstimator.isConverged()) {
+    const result = await this.#simulateInner(builder.build());
+    gasEstimator.update(result);
+    builder.setGasSettings(gasEstimator.proposeGasSettings());
+  }
+
+  return result;
+}
+
+async #ensureFunctionAbis(builder: UserRequestBuilder): void {
+  for (const call of builder.calls) {
+    // User can call simulate without the artifact if they have the function ABI
+    if (!call.functionAbi) {
+      // If the user provides the contract artifact, we don't need to ask the PXE
+      if (!call.contractArtifact) {
+        const contractArtifact = await this.getContractArtifact(call.contractInstance.contractClassId);
+        call.setContractArtifact(contractArtifact);
+      }
+      const functionAbi = findFunctionAbi(call.contractArtifact, call.functionName);
+      call.setFunctionAbi(functionAbi);
+    }
+  }
+}
+
+// helpers somewhere
+
+function decodeSimulatedTx(simulatedTx: SimulatedTx, functionAbi: FunctionAbi): DecodedReturn | [] {
+  const rawReturnValues =
+    functionAbi.functionType == FunctionType.PRIVATE
+      ? simulatedTx.privateReturnValues?.nested?.[0].values
+      : simulatedTx.publicOutput?.publicReturnValues?.[0].values;
+
+  return rawReturnValues ? decodeReturnValues(functionAbi.returnTypes, rawReturnValues) : [];
+}
+```
+
+### BaseWallet.read
+
+Like `simulate`, but without the gas estimation.
+
+```ts
+async read(userRequest: UserRequest): DecodedReturn | [] {
```

Comment:
Indeed. We could scope it to only support static calls, or just nuke it in favor of `simulate`.

Reply:
@spalladino I'm missing where `read` would require the wallet to produce an authwit. User could just use the "signerless" wallet and set `from` on their request?

Reply:
Hmm but doesn't the private kernel init check that msg.sender on the first call of the private stack is zero?

Reply:
Is the "from" value not mostly bogus atm? Since the first call is not what you are actually calling. Otherwise the first authwit at least should be possible to skip with the simulation.